pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,821 @@
+	return ref, nil
+}
+
+func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate, error) {
+
+	update := &buildUpdate{}
+
+	// Set the output Docker image reference.
+	ref, err := bc.resolveOutputDockerImageReference(build)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonInvalidOutputReference)
+		update.setMessage(buildapi.StatusMessageInvalidOutputRef)
+		return update, err
+	}
+
+	// Create the build pod spec
+	buildPod, err := bc.createPodSpec(build, ref)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPodSpec)
+		update.setMessage(buildapi.StatusMessageCannotCreateBuildPodSpec)
+		return update, err
+	}
+
+	glog.V(4).Infof("Pod %s/%s for build %s is about to be created", build.Namespace, buildPod.Name, buildDesc(build))
+	if _, err := bc.podClient.Pods(build.Namespace).Create(buildPod); err != nil {
+		if errors.IsAlreadyExists(err) {
+			bc.recorder.Eventf(build, kapi.EventTypeWarning, "FailedCreate", "Pod already exists: %s/%s", buildPod.Namespace, buildPod.Name)
+			glog.V(4).Infof("Build pod %s/%s for build %s already exists", build.Namespace, buildPod.Name, buildDesc(build))
+
+			// If the existing pod was created before this build, switch to the Error state.
+			existingPod, err := bc.podClient.Pods(build.Namespace).Get(buildPod.Name, metav1.GetOptions{})
+			if err == nil && existingPod.CreationTimestamp.Before(build.CreationTimestamp) {
+				update.setPhase(buildapi.BuildPhaseError)
+				update.setReason(buildapi.StatusReasonBuildPodExists)
+				update.setMessage(buildapi.StatusMessageBuildPodExists)
+				return update, nil
+			}
+			return nil, nil
+		}
+		// Log an event if the pod is not created (most likely due to quota denial).
+		bc.recorder.Eventf(build, kapi.EventTypeWarning, "FailedCreate", "Error creating: %v", err)
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPod)
+		update.setMessage(buildapi.StatusMessageCannotCreateBuildPod)
+		return update, fmt.Errorf("failed to create build pod: %v", err)
+	}
+	glog.V(4).Infof("Created pod %s/%s for build %s", build.Namespace, buildPod.Name, buildDesc(build))
+	update.setPodNameAnnotation(buildPod.Name)
+	update.setPhase(buildapi.BuildPhasePending)
+	update.setReason("")
+	update.setMessage("")
+	update.setOutputRef(ref)
+	return update, nil
+}
+
+func (bc *BuildController) handleActiveBuild(build *buildapi.Build, willBeDropped bool, pod *kapi.Pod, podErr error) (*buildUpdate, error) {
+	switch {
+	case errors.IsNotFound(podErr):
+		// If about to stop retrying the build, fail if necessary
+		if willBeDropped {
+			return bc.handlePodNotFound(build)
+		}
+		// Otherwise, keep retrying
+		return nil, podErr
+	case podErr != nil:
+		return nil, fmt.Errorf("could not retrieve pod for build %s: %v", buildDesc(build), podErr)
+	}
+
+	update := &buildUpdate{}
+	switch pod.Status.Phase {
+	case kapi.PodPending:
+		if build.Status.Phase != buildapi.BuildPhasePending {
+			update.setReason("")
+			update.setMessage("")
```

Comment:
similar issue here.  if the build pod has already failed and updated the build by the time you see this event, you're going to overwrite the failed message/reason.